wallet/hd_wallet.go

```diff
@@ -0,0 +1,299 @@
+	OriginIndex = slip10.FirstHardenedIndex + MagicIndex
+
+	// KeyStartIndex is the first hardened key index to which the key index is
+	// added to derive a key.
+	KeyStartIndex = slip10.FirstHardenedIndex
+)
+
+type HDWallet struct {
+	version uint32
+	name    string
+	keyRing *HDKeyRing
+
+	// node is the node from which the cryptographic keys are generated. This is
+	// not the master node. This is a node derived from the master. Its
+	// derivation index is constant (see OriginIndex). This node is referred as
+	// "wallet node".
+	node *slip10.Node
+}
+
+// NewHDWallet creates a wallet with auto-generated mnemonic. This is useful to
+// create a brand new wallet, without having to take care of the mnemonic
+// generation.
+// The generated mnemonic is returned alongside the created wallet.
+func NewHDWallet(name string) (*HDWallet, string, error) {
+	mnemonic, err := NewMnemonic()
+	if err != nil {
+		return nil, "", err
+	}
+
+	w, err := ImportHDWallet(name, mnemonic)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return w, mnemonic, err
+}
+
+// ImportHDWallet creates a wallet based on the mnemonic in input. This is
+// useful import or retrieve a wallet.
+func ImportHDWallet(name, mnemonic string) (*HDWallet, error) {
+	if !bip39.IsMnemonicValid(mnemonic) {
+		return nil, ErrInvalidMnemonic
+	}
+
+	walletNode, err := deriveWalletNodeFromMnemonic(mnemonic)
+	if err != nil {
+		return nil, err
+	}
+
+	return &HDWallet{
+		version: 1,
+		name:    name,
+		keyRing: NewHDKeyRing(),
+		node:    walletNode,
+	}, nil
+}
+
+func (w *HDWallet) Version() uint32 {
+	return w.version
+}
+
+func (w *HDWallet) Name() string {
+	return w.name
+}
+
+// DescribePublicKey returns all the information associated to a public key. The
+// private key is not returned.
+func (w *HDWallet) DescribePublicKey(pubKey string) (PublicKey, error) {
+	keyPair, ok := w.keyRing.FindPair(pubKey)
+	if !ok {
+		return nil, ErrPubKeyDoesNotExist
+	}
+
+	publicKey := keyPair.ToPublicKey()
+	return &publicKey, nil
+}
+
+// ListPublicKeys lists the public keys with their information. The private keys
+// are not returned.
+func (w *HDWallet) ListPublicKeys() []PublicKey {
+	originalKeys := w.keyRing.ListKeyPairs()
+	keys := make([]PublicKey, len(originalKeys))
+	for i, key := range originalKeys {
+		publicKey := key.ToPublicKey()
+		keys[i] = &publicKey
+	}
+	return keys
+}
+
+// ListKeyPairs lists the key pairs. Be careful, it contains the private key.
+func (w *HDWallet) ListKeyPairs() []KeyPair {
+	originalKeys := w.keyRing.ListKeyPairs()
+	keys := make([]KeyPair, len(originalKeys))
+	for i, key := range originalKeys {
+		keys[i] = key.DeepCopy()
+	}
+	return keys
+}
+
+// GenerateKeyPair generates a new key pair from a node, that is derived from
+// the wallet node.
+func (w *HDWallet) GenerateKeyPair() (KeyPair, error) {
+	nextIndex := w.keyRing.NextIndex()
+	childNode, err := w.node.Derive(KeyStartIndex + nextIndex)
```

Comment:
Shouldn't this be OriginIndex+ nextIndex?

Reply:
If not it means that the first key you generate will be before the one you create when creating the wallet, and you would end up with a collision.
When creating the wallet you use -> OriginIndex which is KeyStartIndex + Magic.
But here you use KeyStartIndex + nextIndex, I imagine at some point nextIndex would be equal to nextIndex, and then you get a collision?

Reply:
Nice catch, you're right ! This is a left over. I thought I removed it...